core/src/services/pcloud/core.rs

```diff
@@ -0,0 +1,445 @@
+            self.endpoint,
+            percent_encode_path(&path),
+            self.username,
+            self.password
+        );
+
+        let req = Request::get(url);
+
+        // set body
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        let resp = self.send(req).await?;
+
+        let status = resp.status();
+        match status {
+            StatusCode::OK => {
+                let bs = resp.into_body().bytes().await?;
+                let resp: GetFileLinkResponse =
+                    serde_json::from_slice(&bs).map_err(new_json_deserialize_error)?;
+                let result = resp.result;
+                if result == 2010 || result == 2055 || result == 2002 {
+                    return Err(Error::new(ErrorKind::NotFound, &format!("{resp:?}")));
+                }
+                if result != 0 {
+                    return Err(Error::new(ErrorKind::Unexpected, &format!("{resp:?}")));
+                }
+
+                if let Some(hosts) = resp.hosts {
+                    if let Some(path) = resp.path {
+                        if !hosts.is_empty() {
+                            return Ok(format!("https://{}{}", hosts[0], path));
+                        }
+                    }
+                }
+                Err(Error::new(ErrorKind::Unexpected, "hosts is empty"))
+            }
+            _ => Err(parse_error(resp).await?),
+        }
+    }
+
+    pub async fn download(&self, url: &str) -> Result<Response<IncomingAsyncBody>> {
+        let req = Request::get(url);
+
+        // set body
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.send(req).await
+    }
+
+    pub async fn ensure_dir_exists(&self, path: &str) -> Result<()> {
+        let path = build_abs_path(&self.root, path);
+
+        let paths = path.split('/').collect::<Vec<&str>>();
+
+        for i in 0..paths.len() - 1 {
+            let path = paths[..i + 1].join("/");
+            let resp = self.create_folder_if_not_exists(&path).await?;
+
+            let status = resp.status();
+
+            match status {
+                StatusCode::OK => {
+                    let bs = resp.into_body().bytes().await?;
+                    let resp: PcloudError =
+                        serde_json::from_slice(&bs).map_err(new_json_deserialize_error)?;
+                    let result = resp.result;
+                    if result == 2010 || result == 2055 || result == 2002 {
+                        return Err(Error::new(ErrorKind::NotFound, &format!("{resp:?}")));
+                    }
+                    if result != 0 {
+                        return Err(Error::new(ErrorKind::Unexpected, &format!("{resp:?}")));
+                    }
+
+                    if result != 0 {
+                        return Err(Error::new(ErrorKind::Unexpected, &format!("{resp:?}")));
+                    }
```

Comment:
Seems we never use status code or headers in `Response<IncomingAsyncBody>`. How about implementing `parse_error(PcloudError) -> Error` instead? So we don't need to repeat the same logic.

Reply:
The result judgment of each operation is not exactly the same, so it is not very easy to extract a function

Reply:
> The result field judgment of each operation is not exactly the same, so it is not very easy to extract a function

Reply:
> The result field judgment of each operation is not exactly the same, so it is not very easy to extract a function

Wow, just know about this..